File: backend/consequence_engine/propagation_test.py
"""
Property test for the server-side CPE. Run from repo root:
    python -m backend.consequence_engine.propagation_test
Mirrors the JS harness (web/src/lib/propagation.test.mjs) to confirm parity.
"""

from datetime import datetime, timezone, timedelta

from backend.consequence_engine import propagation as P

passed = failed = 0


def ok(name, cond):
    global passed, failed
    if cond:
        passed += 1
        print(f"  ok  {name}")
    else:
        failed += 1
        print(f"  XX  {name}")


NOW = datetime(2026, 6, 18, tzinfo=timezone.utc)
NOW_MS = NOW.timestamp() * 1000.0


def iso_ago(h):
    return (NOW - timedelta(hours=h)).isoformat()


def chain(*types):
    return [{"step": i, "type": t, "content": f"n{i}"} for i, t in enumerate(types)]


def mk(id, **o):
    return {
        "id": id,
        "category": o.get("category", "conflict"),
        "canonical_title": o.get("title", f"Event {id}"),
        "importance_score": o.get("importance", 80),
        "first_detected_at": o.get("at", iso_ago(2)),
        "geographic_relevance": o.get("geography", ["Red Sea"]),
        "consequence_map": {
            "consequence_chain": o.get("chain", chain("VERIFIED FACT", "INFERRED MECHANISM", "SPECULATIVE EFFECT")),
            "sources_analyzed": o.get("sources", ["s1", "s2", "s3", "s4"]),
            "disputed_points": o.get("disputed", []),
            "direct_impact": o.get("direct", [{"sector": "Shipping & Logistics", "severity": "critical"}]),
            "indirect_impact": o.get("indirect", [{"sector": "Consumer Prices", "severity": "medium"}]),
        },
    }


def sec(model, key):
    for s in model["sectors"]:
        if s["key"] == key:
            return s["score"]
    return 0


def cem(events, edges=None):
    return P.compute_exposure_model(events, edges or [], NOW_MS)


# 1. determinism
m1 = cem([mk("A"), mk("B", direct=[{"sector": "Shipping", "severity": "high"}])], [{"source": "A", "target": "B", "weight": 0.7}])
m2 = cem([mk("A"), mk("B", direct=[{"sector": "Shipping", "severity": "high"}])], [{"source": "A", "target": "B", "weight": 0.7}])
ok("deterministic", m1 == m2)

# 2. bounded
m = cem([mk("A", importance=100), mk("B", importance=95)], [{"source": "A", "target": "B", "weight": 1}])
allx = m["sectors"] + m["regions"]
ok("bounded [0,100]", all(0 <= x["score"] <= 100 for x in allx))
ok("integers", all(isinstance(x["score"], int) for x in allx))

# 3. attribution
m = cem([mk("A"), mk("B"), mk("C")])
top = m["sectors"][0]
ok("drivers exist", len(top["drivers"]) > 0)
ok("driver pct sum <= 100", sum(d["pct"] for d in top["drivers"]) <= 100)

# 4. importance monotonicity
ok("higher importance => more", sec(cem([mk("A", importance=95)]), "shipping") > sec(cem([mk("A", importance=50)]), "shipping"))

# 5. severity monotonicity
crit = sec(cem([mk("A", direct=[{"sector": "Energy", "severity": "critical"}])]), "energy")
med = sec(cem([mk("A", direct=[{"sector": "Energy", "severity": "medium"}])]), "energy")
ok("critical > medium", crit > med)

# 6. mitigating lowers exposure
base = [mk("A", direct=[{"sector": "Energy", "severity": "critical"}])]
agg = sec(cem(base + [mk("B", direct=[{"sector": "Energy", "severity": "critical"}])]), "energy")
mit = sec(cem(base + [mk("B", direct=[{"sector": "Energy", "severity": "critical", "direction": "mitigating"}])]), "energy")
ok("mitigating < aggravating", mit < agg)

# 7. entity resolution
m = cem([
    mk("A", direct=[{"sector": "Shipping & Logistics", "severity": "high"}], indirect=[]),
    mk("B", direct=[{"sector": "Shipping", "severity": "high"}], indirect=[]),
])
ok("alias collapse", len([s for s in m["sectors"] if s["key"] == "shipping"]) == 1)

# 8. temporal decay
ok("stale < fresh", sec(cem([mk("A", at=iso_ago(24 * 30))]), "shipping") < sec(cem([mk("A", at=iso_ago(1))]), "shipping"))

# 9. corroboration + disputes
ok("more sources => more", sec(cem([mk("A", sources=["1", "2", "3", "4", "5", "6"])]), "shipping") > sec(cem([mk("A", sources=["1"])]), "shipping"))
ok("disputes => less", sec(cem([mk("A", disputed=["x", "y"])]), "shipping") < sec(cem([mk("A", disputed=[])]), "shipping"))

# 10. directed vs undirected
A = mk("A", direct=[{"sector": "Energy", "severity": "critical"}], indirect=[])
B = mk("B", direct=[{"sector": "Banking", "severity": "low"}], indirect=[], importance=20)
di = sec(cem([A, B], [{"source": "A", "target": "B", "weight": 0.9, "directed": True}]), "banking")
un = sec(cem([A, B], [{"source": "A", "target": "B", "weight": 0.9}]), "banking")
ok("undirected >= directed", un >= di)
ok("directed still amplifies", di > 0)

# 11. profile exposure
model = cem([mk("A"), mk("B", direct=[{"sector": "Energy", "severity": "critical"}], geography=["Europe"])])
pe = P.profile_exposure({"sectors": ["Energy"], "regions": ["Europe"]}, model)
ok("profile bounded", 0 <= pe["score"] <= 100)
ok("unmatched => 0", P.profile_exposure({"sectors": ["Nonexistent"]}, model)["score"] == 0)

# 12. no secret params leaked in meta
ok("meta has version", model["meta"]["version"] == P.ENGINE_VERSION)
ok("meta hides LAMBDA/K", "LAMBDA" not in model["meta"] and "K" not in model["meta"])

# 13. per-event exposure heat
m13 = cem([mk("A", importance=95), mk("B", importance=40)])
ok("event_scores present", m13["event_scores"].get("A") is not None and m13["event_scores"].get("B") is not None)
ok("event_scores bounded", all(0 <= s <= 100 for s in m13["event_scores"].values()))
ok("higher-driving event hotter", m13["event_scores"]["A"] > m13["event_scores"]["B"])

# 14. traffic disruption feeds the CPE
ev = [mk("A", direct=[{"sector": "Energy", "severity": "low"}], indirect=[], geography=["Red Sea"])]
base = P.compute_exposure_model(ev, [], NOW_MS)
wt = P.compute_exposure_model(ev, [], NOW_MS, {"A": {"vessels": 200, "aircraft": 50}})
ok("no shipping without traffic", sec(base, "shipping") == 0)
ok("traffic raises shipping", sec(wt, "shipping") > 0)
ok("traffic raises aviation", sec(wt, "aviation") > 0)

# 15. market stress amplifies stressed sectors
ev_m = [mk("A", direct=[{"sector": "Energy", "severity": "medium"}], indirect=[])]
base_e = sec(P.compute_exposure_model(ev_m, [], NOW_MS), "energy")
boost_e = sec(P.compute_exposure_model(ev_m, [], NOW_MS, None, {"Energy": 0.9}), "energy")
ok("market stress raises stressed sector", boost_e > base_e)
mkt_only = P.compute_exposure_model([], [], NOW_MS, None, {"Commodities": 0.8})
ok("market-only sector appears", sec(mkt_only, "commodities") > 0)

# 16. combine_stress merges independent stress sources (max per sector, clamped)
cs = P.combine_stress({"Energy": 0.5}, {"Energy": 0.7, "Shipping & Logistics": 0.3}, {"Aerospace": 0.9})
ok("combine takes max per sector", cs["Energy"] == 0.7)
ok("combine unions distinct sectors", cs["Shipping & Logistics"] == 0.3 and cs["Aerospace"] == 0.9)
ok("combine clamps to [0,1]", P.combine_stress({"X": 5, "Y": -2}) == {"X": 1.0})
ok("combine ignores None + bad values", P.combine_stress(None, {"Z": "bad"}) == {})

# 17. combined chokepoint + space-weather stress amplifies their sectors via the CPE
ev_c = [mk("A", direct=[{"sector": "Energy", "severity": "low"}], indirect=[])]
combined = P.combine_stress({"Shipping & Logistics": 0.8}, {"Aerospace": 0.9})
mc = P.compute_exposure_model(ev_c, [], NOW_MS, None, combined)
ok("chokepoint stress surfaces Shipping", sec(mc, "shipping") > 0)
ok("space-weather stress surfaces Aerospace", sec(mc, "aerospace") > 0)

# 18. cross-feed corroboration boosts a corroborated event's exposure
ev_co = [mk("A", importance=40, direct=[{"sector": "Energy", "severity": "low"}], indirect=[])]
base_co = sec(P.compute_exposure_model(ev_co, [], NOW_MS), "energy")
boost_co = sec(P.compute_exposure_model(ev_co, [], NOW_MS, corroboration={"A": {"index": 1.0}}), "energy")
ok("corroboration boosts exposure", boost_co > base_co)
ok("no corroboration ⇒ unchanged", sec(P.compute_exposure_model(ev_co, [], NOW_MS, corroboration=None), "energy") == base_co)
ok("zero index ⇒ unchanged", sec(P.compute_exposure_model(ev_co, [], NOW_MS, corroboration={"A": {"index": 0.0}}), "energy") == base_co)

print(f"\nServer CPE v{P.ENGINE_VERSION}: {passed} passed, {failed} failed")
raise SystemExit(1 if failed else 0)

File: web/src/lib/propagation.test.mjs
// Property + golden test harness for the Consequence Propagation Engine (CPE).
// No test framework needed — run with:  node web/src/lib/propagation.test.mjs
// Proves the guarantees claimed in the plan: bounded 0–100, deterministic,
// monotonic in severity/importance, attribution well-formed, Katz converges,
// mitigating edges lower net exposure, entity resolution, temporal decay,
// directed propagation flows downstream only.

import { computeExposureModel, CPE } from "./propagation.js";
import { profileExposure } from "./exposureProfile.js";

let passed = 0, failed = 0;
const ok = (name, cond) => { if (cond) { passed++; console.log(`  ✓ ${name}`); } else { failed++; console.error(`  ✗ ${name}`); } };

const NOW = new Date("2026-06-18T00:00:00Z").getTime();
const HOUR = 3_600_000;
const isoAgo = (h) => new Date(NOW - h * HOUR).toISOString();

const chain = (...types) => types.map((t, i) => ({ step: i, type: t, content: `n${i}` }));

function mkEvent(id, over = {}) {
  return {
    id,
    category: over.category || "conflict",
    canonical_title: over.title || `Event ${id}`,
    importance_score: over.importance ?? 80,
    first_detected_at: over.at ?? isoAgo(2),
    geography: over.geography || ["Red Sea"],
    consequence_map: {
      consequence_chain: over.chain || chain("VERIFIED FACT", "INFERRED MECHANISM", "SPECULATIVE EFFECT"),
      sources_analyzed: over.sources || ["s1", "s2", "s3", "s4"],
      disputed_points: over.disputed || [],
      direct_impact: over.direct || [{ sector: "Shipping & Logistics", severity: "critical" }],
      indirect_impact: over.indirect || [{ sector: "Consumer Prices", severity: "medium" }],
    },
  };
}

const sectorScore = (model, key) => model.sectors.find((s) => s.key === key)?.score ?? 0;
const opts = { now: NOW };

// ── 1. Determinism ────────────────────────────────────────────────────────────
{
  const events = [mkEvent("A"), mkEvent("B", { direct: [{ sector: "Shipping", severity: "high" }] })];
  const edges = [{ source: "A", target: "B", weight: 0.7 }];
  const m1 = computeExposureModel(events, edges, opts);
  const m2 = computeExposureModel(events, edges, opts);
  ok("deterministic across runs", JSON.stringify(m1) === JSON.stringify(m2));
}

// ── 2. Bounded 0–100 integers ──────────────────────────────────────────────────
{
  const events = [mkEvent("A", { importance: 100 }), mkEvent("B", { importance: 95 })];
  const m = computeExposureModel(events, [{ source: "A", target: "B", weight: 1 }], opts);
  const all = [...m.sectors, ...m.regions];
  ok("all scores within [0,100]", all.every((x) => x.score >= 0 && x.score <= 100));
  ok("all scores are integers", all.every((x) => Number.isInteger(x.score)));
  ok("pressure within [0,100]", m.pressure >= 0 && m.pressure <= 100);
}

// ── 3. Attribution well-formed ─────────────────────────────────────────────────
{
  const events = [mkEvent("A"), mkEvent("B"), mkEvent("C")];
  const m = computeExposureModel(events, [], opts);
  const top = m.sectors[0];
  ok("top entity has drivers", top.drivers.length > 0);
  ok("driver pct in [0,100]", top.drivers.every((d) => d.pct >= 0 && d.pct <= 100));
  ok("drivers sorted descending by pct", top.drivers.every((d, i, a) => i === 0 || a[i - 1].pct >= d.pct));
  ok("driver pct sum ≤ 100 (top-3 share)", top.drivers.reduce((s, d) => s + d.pct, 0) <= 100);
}

// ── 4. Monotonicity in importance ──────────────────────────────────────────────
{
  const low = computeExposureModel([mkEvent("A", { importance: 50 })], [], opts);
  const high = computeExposureModel([mkEvent("A", { importance: 95 })], [], opts);
  ok("higher importance ⇒ ≥ exposure", sectorScore(high, "shipping") >= sectorScore(low, "shipping"));
  ok("higher importance ⇒ strictly more here", sectorScore(high, "shipping") > sectorScore(low, "shipping"));
}

// ── 5. Severity monotonicity ───────────────────────────────────────────────────
{
  const med = computeExposureModel([mkEvent("A", { direct: [{ sector: "Energy", severity: "medium" }] })], [], opts);
  const crit = computeExposureModel([mkEvent("A", { direct: [{ sector: "Energy", severity: "critical" }] })], [], opts);
  ok("critical severity > medium severity", sectorScore(crit, "energy") > sectorScore(med, "energy"));
}

// ── 6. Mitigating impacts lower net exposure ───────────────────────────────────
{
  const base = [mkEvent("A", { direct: [{ sector: "Energy", severity: "critical" }] })];
  const aggravate = [...base, mkEvent("B", { direct: [{ sector: "Energy", severity: "critical" }] })];
  const mitigate = [...base, mkEvent("B", { direct: [{ sector: "Energy", severity: "critical", direction: "mitigating" }] })];
  const sA = sectorScore(computeExposureModel(aggravate, [], opts), "energy");
  const sM = sectorScore(computeExposureModel(mitigate, [], opts), "energy");
  ok("mitigating event lowers exposure vs aggravating", sM < sA);
}

// ── 7. Entity resolution (alias collapse) ──────────────────────────────────────
{
  const events = [
    mkEvent("A", { direct: [{ sector: "Shipping & Logistics", severity: "high" }], indirect: [] }),
    mkEvent("B", { direct: [{ sector: "Shipping", severity: "high" }], indirect: [] }),
  ];
  const m = computeExposureModel(events, [], opts);
  const shippingEntries = m.sectors.filter((s) => s.key === "shipping");
  ok("'Shipping & Logistics' + 'Shipping' collapse to one entity", shippingEntries.length === 1);
}

// ── 8. Temporal decay ──────────────────────────────────────────────────────────
{
  const fresh = computeExposureModel([mkEvent("A", { at: isoAgo(1) })], [], opts);
  const stale = computeExposureModel([mkEvent("A", { at: isoAgo(24 * 30) })], [], opts); // 30 days old
  ok("stale event contributes less than fresh", sectorScore(stale, "shipping") < sectorScore(fresh, "shipping"));
}

// ── 9. Corroboration & disputes affect confidence ──────────────────────────────
{
  const wellSourced = computeExposureModel([mkEvent("A", { sources: ["1", "2", "3", "4", "5", "6"] })], [], opts);
  const thin = computeExposureModel([mkEvent("A", { sources: ["1"] })], [], opts);
  ok("more sources ⇒ higher exposure (corroboration)", sectorScore(wellSourced, "shipping") > sectorScore(thin, "shipping"));

  const undisputed = computeExposureModel([mkEvent("A", { disputed: [] })], [], opts);
  const disputed = computeExposureModel([mkEvent("A", { disputed: ["x", "y"] })], [], opts);
  ok("disputed points ⇒ lower exposure", sectorScore(disputed, "shipping") < sectorScore(undisputed, "shipping"));
}

// ── 10. Directed propagation flows downstream only ─────────────────────────────
{
  // A (cause) → B (effect). A's sector is Energy, B's sector is Tech.
  const A = mkEvent("A", { direct: [{ sector: "Energy", severity: "critical" }], indirect: [] });
  const B = mkEvent("B", { direct: [{ sector: "Banking", severity: "low" }], indirect: [], importance: 20 });
  const directed = computeExposureModel([A, B], [{ source: "A", target: "B", weight: 0.9, directed: true }], opts);
  const undirected = computeExposureModel([A, B], [{ source: "A", target: "B", weight: 0.9 }], opts);
  // Banking (B's sector) should get MORE amplification from A in the undirected case
  // (A→B and B→A) than in the directed case (A→B only).
  ok("undirected amplifies Banking ≥ directed", sectorScore(undirected, "banking") >= sectorScore(directed, "banking"));
  ok("directed case still amplifies effect's sector", sectorScore(directed, "banking") > 0);
}

// ── 11. Profile exposure blends matched dimensions ─────────────────────────────
{
  const events = [mkEvent("A"), mkEvent("B", { direct: [{ sector: "Energy", severity: "critical" }], geography: ["Europe"] })];
  const model = computeExposureModel(events, [], opts);
  const pe = profileExposure({ sectors: ["Energy"], regions: ["Europe"] }, model);
  ok("profile exposure within [0,100]", pe.score >= 0 && pe.score <= 100);
  ok("profile exposure returns drivers", Array.isArray(pe.drivers));
  const empty = profileExposure({ sectors: ["Nonexistent Sector"] }, model);
  ok("unmatched profile ⇒ 0", empty.score === 0 && empty.drivers.length === 0);
}

// ── 12. Converges well under MAX_ITERS on a dense graph ────────────────────────
{
  const events = Array.from({ length: 12 }, (_, i) => mkEvent(`E${i}`, { importance: 60 + i }));
  const edges = [];
  for (let i = 0; i < 12; i++) for (let j = 0; j < 12; j++) if (i !== j) edges.push({ source: `E${i}`, target: `E${j}`, weight: 0.4 });
  const t0 = Date.now();
  const m = computeExposureModel(events, edges, opts);
  const dt = Date.now() - t0;
  ok("dense-graph run completes < 500ms", dt < 500);
  ok("dense-graph scores bounded", [...m.sectors, ...m.regions].every((x) => x.score >= 0 && x.score <= 100));
}

// ── 13. Per-event exposure heat (eventScores) ──────────────────────────────────
{
  const m = computeExposureModel([mkEvent("A", { importance: 95 }), mkEvent("B", { importance: 40 })], [], opts);
  ok("eventScores present for each event", m.eventScores.A != null && m.eventScores.B != null);
  ok("eventScores bounded [0,100]", Object.values(m.eventScores).every((s) => s >= 0 && s <= 100));
  ok("higher-driving event has higher heat", m.eventScores.A > m.eventScores.B);
}

// ── 14. Traffic disruption feeds the CPE ───────────────────────────────────────
{
  const ev = [mkEvent("A", { direct: [{ sector: "Energy", severity: "low" }], indirect: [], geography: ["Red Sea"] })];
  const base = computeExposureModel(ev, [], opts);
  const withTraffic = computeExposureModel(ev, [], { now: NOW, trafficByEvent: { A: { vessels: 200, aircraft: 50 } } });
  ok("no shipping exposure without traffic", sectorScore(base, "shipping") === 0);
  ok("heavy nearby traffic raises shipping exposure", sectorScore(withTraffic, "shipping") > 0);
  ok("nearby flights raise aviation exposure", sectorScore(withTraffic, "aviation") > 0);
}

console.log(`\nCPE v${CPE.VERSION}: ${passed} passed, ${failed} failed`);
process.exit(failed ? 1 : 0);
